### PR TITLE
Fix test_extract_class_dict for Python 3.13 beta 1

### PR DESCRIPTION
**Pull request by frenzymadness**
Resolves: https://github.com/cloudpipe/cloudpickle/issues/533

### CODE REVIEW

**Comment by ogrisel** [comment]
Thanks for the PR. Ideally we would need to update the CI. Do you know if that can be easily achieved with GitHub Actions?

**Comment by frenzymadness** [comment]
Python 3.13 is already available in setup-python action, see: https://github.com/actions/python-versions/blob/main/versions-manifest.json

So it should be easy to add 3.13 to the CI here.

**Comment by ogrisel** [comment]
The remaining failures seems to be all related to @tomMoral's tests for deterministic pickling implemented in #524.

I had a quick look but it's not easy to understand why cloudpickle with Python 3.13 no longer leads to deterministic pickling.

**Comment by ogrisel** [comment]
More specifically, for `test_deterministic_dynamic_class_attr_ordering_for_chained_pickling`, the first difference is the lack of memoization of the `"CloudPickleTest.test_deterministic_dynamic_class_attr_ordering_for_chained_pickling.<locals>.A.__init__"` string, which looks like the fully qualified name of the `__init__` method of the class being pickled in the test.

https://github.com/cloudpipe/cloudpickle/actions/runs/10183782743/job/28169551899?pr=534#step:6:168

**Comment by ogrisel** [comment]
Maybe this is the cause: https://github.com/cloudpipe/cloudpickle/pull/524/files#r1399015881

**Comment by ogrisel** [comment]
Ok d6696c2 did the trick! I will update the changelog and let @tomMoral some time to review this change.

**Comment by QuLogic** [comment]
Python 3.13 final release is in a week, and this appears to be passing tests; is there any thing remaining to complete this (other than review, I suppose)?

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -707,7 +707,7 @@ def _function_getstate(func):
         # Hack to circumvent non-predictable memoization caused by string interning.
         # See the inline comment in _class_setstate for details.
         "__name__": "".join(func.__name__),
-        "__qualname__": func.__qualname__,
+        "__qualname__": "".join(func.__qualname__),
```
Comment (tomMoral) [inline cloudpickle/cloudpickle.py:710]: It was previously not interned but now it is without rebuilding the string?
Not super clear why this behavior changed, I could not find any lead in the python3.13 release note.

But if it does the trick, this seems fine like this, this should not be too costly to do.

**Review by tomMoral — APPROVED** [review]
LGTM! Thanks @frenzymadness and @ogrisel for the debug of this part.

Note that python 3.13 now have a `sys.isinterned` which might be useful to better investigate this issue.

**Review by ogrisel — APPROVED** [review]

**Comment by ogrisel** [comment]
After triggering the tests again with CPython 3.13.0 we get new failures on macOS. Will investigate.

**Comment by ogrisel** [comment]
The `__firstlineno__` is present in the first pickle string but missing from the second in:

```python
            class A:
                """Simple class definition"""

                pass

            A_dump = w.run(cloudpickle.dumps, A)
            check_deterministic_pickle(A_dump, cloudpickle.dumps(A))
```

**Comment by pierreglaser** [comment]
I did a bit more digging. In Python 3.13, [setting the `__module__` attribute of some class has the side effect of deleting its `__firstlineno__` attribute](https://github.com/python/cpython/blob/73c152b346a18ed8308e469bdd232698e6cd3a63/Objects/typeobject.c#L1353-L1356):

```python

In [1]: class A:
      ...:     pass
      ...:

In [2]: A.__firstlineno__
Out[2]: 1

In [3]: A.__module__ = "anymodule"

In [4]: A.__firstlineno__
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 A.__firstlineno__

AttributeError: type object 'A' has no attribute '__firstlineno__'
```

Note that this snippet does not involve `cloudpickle`.

The cloudpickle reproducibility tests fail because of the folllowing sequence of events:
- a class is created, and pickled using `cloudpickle`, generating a first pickle string. Since it's `__module__` is left untouched in the process, the class has a `__firstlineno__` attribute, which is pickled as part of the class's state by `cloudpickle`
- at unpickling time, `cloudpickle` first restores the class's `__firstlineno__` attribute,  and **then** `__module__`, thus deleting the restored `__firstlineno__` attribute. 
- this depickled class is then repickled another time, yielding a second pickle string, which differs from the first one as it not contain a `__firstlineno__`  attribute in it. Consequently,  the test, which asserts the equality of the two strings, fails.

There is a question to be had about whether the `__firstlineno__` attribute of a depickled dynamically created class has any valuable meaning. But from a user perspective, the least surprising behaviour that cloudpickle can have is probably to restore this argument as it was as unpickling time, which is what I propose we do.

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -1167,6 +1167,13 @@ def _class_setstate(obj, state):
             # Indeed the Pickler's memoizer relies on physical object identity to break
             # cycles in the reference graph of the object being serialized.
             setattr(obj, attrname, attr)
+
+    if sys.version_info >= (3, 13) and "__firstlineno__" in state:
+        # Set the Python 3.13+ only __firstlineno__  attribute one more time, as it
+        # will be automatically deleted by the `setattr(obj, attrname, attr)` call
+        # above when `attrname` is "__firstlineno__".
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:1174]:
```suggestion
        # above when `attrname` is "__firstlineno__". We assume that preserving this
        # information might be important for some users and that it not stale in the
        # context of cloudpickle usage, hence legitimate to propagate. Furthermore it
        # is necessary to do so to keep determinitic chained pickling as tested in
        # test_deterministic_str_interning_for_chained_dynamic_class_pickling.
```

**Review by ogrisel — APPROVED** [review]

File: CHANGES.md [hunk 3]
```diff
@@ -3,7 +3,8 @@
 
 - Some improvements to make cloudpickle more deterministic when pickling
   dynamic functions and classes.
```
Comment (ogrisel) [inline CHANGES.md:5]:
```suggestion
  dynamic functions and classes, in particular with CPython 3.13.
```

**Comment by pierreglaser** [comment]
Merging, thanks @tomMoral  @frenzymadness  @ogrisel!

**Pull request merged**